### DISCUSSION TITLE
must address this feature please! alignment is necessary!

### DISCUSSION THREAD

**Discussion by donrikk** [nagadomi/nunif#142] [Ideas]
hey nagadomi so i have found that the iw3 converter does waht its supposed to do which is fantastic! BUT, its missing one thing that is meant to finalize the 3d conversion properly to make it 1 to 1 to a 3d bluray, ive done multiple comparisons in grey anaglyph comparing bluray 3d copies to iw3 conversion. i found that the iw3 result NEEDS alignment horizontally. the sbs images need to be aligned so that the closest of the foreground object is at zero parallax or if u look at it in grey anaglyph the foreground object if properly aligned should be completely grey with no red or cyan, u basically want to align the sbs images negatively to align the nearest foreground object THEN after thats align is where the .5 convergence should take place, this creates a 1 to 1 bluray clone! its the only thing missing really from the conversion and its objectively needed to get a finalized result.

below are examples with there respective label depicting the difference in the convegence and alignment and how alignment is absolutely needed for a final result, it shouldnt really add any added time to the conversion as its only a horizontal shift in frame maybe a couple more minutes but thats all. i hope im not coming off as rough as i think i am im just really passionate about this project and this is something that would really help ur converter by miles! much love! here below are the examples!

![frame_23  con P05jpg](https://github.com/nagadomi/nunif/assets/106938990/640c52c4-7971-423a-aa8c-1feb8fb9a8ac)
![frame_23 converge 5](https://github.com/nagadomi/nunif/assets/106938990/4fa29919-24d7-48d0-ae2b-59d9624c4895)
![frame_23 with alignment](https://github.com/nagadomi/nunif/assets/106938990/4d43b5a0-5149-4592-86a7-467973c68774)
![frame_23 without alignment](https://github.com/nagadomi/nunif/assets/106938990/df428dc0-c525-4ec0-aaff-d7962d3d648e)

**Comment by donrikk** [comment]
here is a video example of what convergence does to the image vs alignment, its a bit slow of a video so feel free to scrub through! 


https://github.com/nagadomi/nunif/assets/106938990/074e9143-3ef5-4267-bb9e-672b09e1c87c

**Reply by nagadomi** [reply]
Which is alignment you are talking about, the first or the second?
From your other comments, it sounds like convergence in iw3 is alignment you are talking about.

**Reply by donrikk** [reply]
@nagadomi alignment is the first part of the video. and convergence is represented the other half of the video so first thing im doing is showing how this slider bar allows me to manual set the alignment of both frames. while the second half that shows convergence showcases that convergence simply tightens the image to center thus stretching the entire image and only enhancing the pop thats already established. currently in iw3 convergence works the same, 

convergence is needed but only after alignment. 

what alignment does basically is it take the left and right eye and it shifts them to the left and or right horizontally. so in the picture examples u can see what it looks like fresh out of iw3 VS after an alignment, the alignment picture example is where the frames should be aligned. when you are doing the alignment and ur shifting inwards to zero out the parralax of the foreground it in return deepens everything behind the foreground into the screen. this allows only the closest of foreground object to pop out of screen like intended while also properly setting the mid and background where it needs to be in relation to that foreground.

like in the gorilla example the arm is the closest of the body parts of the gorilla to the foreground so with alignment ull see that it sets only the arm protruding out of screen while the rest of the gorilla behind his arm is zero parallax and slowly recedes into the back like its meant too. 

this alignment process should be done before convergence so once the video is aligned then convergence should be applied. 

here is a conversion example of a iw3 conversion being aligned in combine like in the video above. 

https://drive.google.com/file/d/1POPxMLfBpGxdCQX8wQ4ja9jEG8E1FUuE/view?usp=drive_link

**Reply by nagadomi** [reply]
`Convergence Plane` in iw3 is the setting of where in 1-0 (1=most foreground, 0= most background) to be zero parallax point. So it is `alignment` you are talking about.
![convergence](https://github.com/nagadomi/nunif/assets/287255/ca8ce084-e02f-4098-8f4b-726bd6505f60)
Also I think, convergece you are talking about is the ability to correct the different geometric distortions of the left and right cameras. iw3 uses a single image as input, there is no need to do it.

**Reply by donrikk** [reply]
@nagadomi let me whip up a example of iws conversion at 0.5 convergence compared to my aligned version lets see if there is a difference

**Reply by donrikk** [reply]
so they are pretty much identical actually! so i think what u have labeled convergence might acually be alignment. because convegence doesnt act like iw3s convergence option it acts more like alignment. thats my bad! convergence is usually a system where the image is either pinch in the middle or stretched outward towards the sides of the video u can see during convergence that the image gets stretched into position while alignment aligns the frames into position without disrupting the actual image itself.

**Reply by loawizard** [reply]
so convergence is actually allignment then ? :D this is so confusing

**Reply by donrikk** [reply]
@loawizard considering im seeing the same results as alignment in combine when using convergence feature and i dont see the same effect as a regular convergence on combine or my other 3d photo editors. so i do believe the convergence setting in iw3 is more akeen to alignment and convergence combined instead of just convergence. i say that becuase mine is identical to a .5 converge output from iw3 which means the image does go through a alignemnt and convergence process as thats the only way i can achieve that in combine. but its so good to know the program does this lol cuz then i dont have to deal with bouncing between programs XD so thanks @nagadomi for the clarification it was helpful!

**Reply by donrikk** [reply]
@nagadomi i asked ai the difference betwen alignment and convergence and this is what it told me

Stereo Pair Video Alignment:

Stereo pair video alignment refers to the process of ensuring that the left and right views of a stereoscopic pair are correctly aligned spatially. This alignment is crucial for creating a comfortable and accurate 3D viewing experience.
In stereo pair alignment, the goal is to match corresponding points or features in the left and right views so that they appear correctly aligned when viewed through 3D glasses or a stereoscope. Misalignment can lead to visual discomfort or incorrect depth perception.
Convergence:

Convergence, on the other hand, refers to the point where the lines of sight of the viewer intersect. It is the point towards which both eyes converge when looking at an object or scene.
In stereoscopic 3D video, convergence is adjusted to control the perceived depth of the scene. By adjusting convergence, the depth plane (where objects appear to lie in space relative to the screen) can be moved closer or farther away from the viewer.
Proper convergence adjustment is critical to ensure that the 3D effect is comfortable and realistic. Incorrect convergence settings can lead to eyestrain or a distorted perception of depth.
Key Differences:

Purpose: Stereo pair video alignment ensures that corresponding points in left and right views match precisely. Convergence adjustment controls the perceived depth.

Adjustment: Alignment is about spatial alignment of corresponding pixels or features. Convergence is about adjusting the depth plane to match the content being viewed.

Effects: Misalignment affects the visual comfort and accuracy of 3D perception. Incorrect convergence affects the perceived depth and can cause discomfort.

In summary, while both stereo pair video alignment and convergence are crucial for creating a good stereoscopic 3D viewing experience, they serve different purposes: alignment ensures spatial coherence between left and right views, while convergence adjusts the depth perception of the scene. Both factors need to be carefully calibrated to achieve a natural and comfortable 3D effect.

**Comment by kaelsonofkrypto** [comment]
Nagadomi, Donrik is right about the alignment part, if u could somehow implement what he is asking it will be 99 percent perfect, its already at 95 percent there of near perfection but this alignment thing he mentioned is right, i saw some demos of what he was saying and it works. And as always thanks so much for making something really great, what u have made is already great by the way, this would just get it closer to absolute perfection.

**Comment by Ryudoadema** [comment]
This would be a great addition!

**Comment by loawizard** [comment]
hmnn this is interesting.. cool feature.

Also thank you for making IW3 faster with your last updates. AMAZING Work

**Comment by diverswan9** [comment]
Have you tried Convergence plane 0.9 ? That should put it pretty close to your bluray example. Or even 1.0

**Reply by donrikk** [reply]
the problem is is convergence pinches and contracts the image it does not set the background farther back in the video plane. alignment and convergence are fundementally different. convergence pinches and contracts the image depending on if ur doing negative or positive convergence, this will tighten the image to make it more pleasing to the eye BUT it does ZERO aligning of the image, the foreground and background will still be in the same plane of depth that it was without convergence it simply tightens the objects to the middle of the screen so you have less retinal rivalry, alignment shift the entire image left or right within the horizontal plane to line up both sides of the sbs video properly this alignment process does no warping and no contracting or pinching it simply shifts the frames into proper position. so adding 0.9 convergence doesnt do any of what alignment does all it wil do is make the foreground pop out further but it wont change the fact that the background are at zero parallax when they should be set far bac kin the scene. for 3d to properly work the background should be no where near the foreground especially at zero parallax. the only thing that should be at zero parallax in the 3d image is the closest of foreground objects and that inlcudes like say someone lifting there arm in the scene the arm since it is the closest to the foreground should be at zero while the person body should be shifted slightly behind that arm. this is how u get proper pop out. currently everything in the foreground pops out without doing alignment but thats not how its supposed to work, not everything within the foreground of a image is supposed to pop out only the closest of the foreground.

**Reply by diverswan9** [reply]
@donrikk "so adding 0.9 convergence doesnt do any of what alignment does all it wil do is make the foreground pop out further"

0.9 has minimal pop-out. 1.0 has 0 pop-out. I'm referring to iw3. Not another program.

**Reply by donrikk** [reply]
yes in iw3 convergence works the same way. convergence will not set the background farther way only alignment is that acheivable. no amount of convergence will change the position of the background and foreground, it will only reduce retinal rivalry and help the pop out effect look more clear and crisp which is why i suggest it AFTER the alignment, that way u get the image aligned and u get the benefits of a crisper pop out. take a look at the video i posted in the comment below my original post here, if u watch u can see what alignment does compared to convergence. they dont even come close to doing the same thing iw3 or combine or owl or any 3d converter it doesnt matter alignemnt is fundementally different from convergence 😄

**Reply by donrikk** [reply]
@diverswan9 for instance have u ever viewed a real studio sbs conversion? have u ever notice in most of them there is this vertical black bar usually on the right frame on the right side of the right image in the sbs that seems to shrink and expand throughout the playing of the video. that black bar is demonstrating the sbs video aligning in real time frame by frame like what im doing here, only i crop out the black bar completely so it isnt in my final result but some 3d bluray conversion have that black bar. so even studios do this alignment process. it is seperate from convergence but both convergence and alignment work together by the end result. but alignment comes first then convergence.

**Comment by kaelsonofkrypto** [comment]
@nagadomi @donrikk 

Proposed solution: Here is how it could possibly be implemented in iw3.

Step 1: Automatic Horizontal Alignment for SBS Images
•	create a Field name: "Alignment"
•	Options: "Auto" (identifies nearest object and adjusts alignment for zero parallax) or "Manual" (user-defined alignment value)
•	Algorithm: Analyze SBS images, adjust alignment based on chosen option and set nearest object at zero parallax.

Step 2: Integrate Auto-Alignment with Convergence
•	Modify the existing "Converge" field: Add a dropdown selection “Apply auto-alignment before convergence (0.5 or 1)
•	Update algorithm: Reference auto-alignment results during SBS image processing

Thanks for all that you do Nagadomi

**Reply by donrikk** [reply]
good solution right here! love it! if an automatic is possible ide go with that but if all else fails a slider system with horizontal adjustment wit ha 1 frame preview to test the alignment could do wonders as u only have to shift the video once and all the other frames will cohere to it. as the alignment integer will vary depending on depth strength so for instance at 1.45 depth strength in iw3 takes a -10 horizontal alignment but with a video at 2.5 depth strength ull have to shift the images around the -20 to -30 range. so strength effects the amount of alignment thats needed.
